Solutions/0.2-Elliptic-Curve-Math-Solutions.ipynb

```diff
@@ -0,0 +1,272 @@
+    "\n",
+    "`int`: All Scalar operations over SECP251k1 can be performed with python integers and the modulo `%` operator:\n",
+    "* All scalar operations are performed modulo n (`SECP256k1_ORDER`).\n",
+    "* Addition: `a + b % SECP256K1_ORDER`\n",
+    "* Subtraction: `-a = SECP256K1_ORDER - a`\n",
+    "* Multiplication: `a * b % % SECP256K1_ORDER`\n",
+    "* Division (Fermat's little Theorem): `1/b = b**(SECP256K1_ORDER-2) % SECP256K1_ORDER`\n",
+    "\n",
+    "`ECKey`: The Bitcoin Core library provides a private key class which can also perform certain scalar operations.\n",
+    "* Addition: `add`\n",
+    "* Subtraction: `add` (adding complement of scalar)\n",
+    "* Multiplication: `mul`\n",
+    "* Division: `mul` (multiplying with scalar inverse, see Fermat's little theorem) \n",
+    "\n",
+    "![test](../images/ec_math1.jpg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 1.1 Example: Scalar Addition over SECP256K1 order\n",
+    "\n",
+    "Addition can be performed with integer modulo math in python or with the private key class `ECKey`. We can set a `ECKey` object to a certain value, or generate a new private key with the `generate` method.\n",
+    "\n",
+    "In the example below, addition is performed with both integers and the `ECKey` class, and evaluated for equality."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# int() operations\n",
+    "a = random.randrange(1, SECP256K1_ORDER)\n",
+    "b = random.randrange(1, SECP256K1_ORDER)\n",
+    "ab = (a + b) % SECP256K1_ORDER\n",
+    "\n",
+    "# ECkey(operation)\n",
+    "a_key = ECKey()\n",
+    "a_key.set(a.to_bytes(32,'big'), True) \n",
+    "ab_key = a_key.add(b.to_bytes(32,'big'))\n",
+    "\n",
+    "# Ensure equivalency.\n",
+    "print(int().from_bytes(ab_key.get_bytes(),'big') == ab)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 1.2 Example: Scalar Multiplication over SECP256K1 order\n",
+    "\n",
+    "In the example below, multiplication is performed with both integers and the `ECKey` class, and evaluated for equality."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# int() operations\n",
+    "a = random.randrange(1, SECP256K1_ORDER)\n",
+    "b = random.randrange(1, SECP256K1_ORDER)\n",
+    "ab = (a * b) % SECP256K1_ORDER\n",
+    "\n",
+    "# ECkey(operation)\n",
+    "a_key = ECKey()\n",
+    "a_key.set(a.to_bytes(32,'big'), True) \n",
+    "ab_key = a_key.mul(b.to_bytes(32,'big'))\n",
+    "\n",
+    "# Ensure operations are equivalent.\n",
+    "print(int().from_bytes(ab_key.get_bytes(),'big') == ab)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 1.3 _Programming Exercise:_ Distributivity of scalar operations\n",
```

Comment:
section numbering (applies to this whole notebook)

Reply:
fixed